ARCHITECTURE.md

```diff
@@ -0,0 +1,32 @@
+# Architecture
+
+This document describes the high-level architecture of realtime_signs. The goal of this document is to provide a useful
+orientation to a new contributor to the codebase.
+
+## Application description
+
+Realtime Signs is the app that puts subway countdown predictions on the black and yellow Daktronics LED signs in and around stations. It does this by continually (about 1/sec) downloading GTFS-rt data and sending HTTP POST requests as necessary to the ARINC head end server.
+
+## Process-level architecture
+
+The smallest granularity with which we can address signs in a station is called the "zone". Each station can have up to six zones: north, south, east, west, center, and mezzanine. Each zone throughout the system has its own dedicated Elixir process, a `Signs.Realtime` GenServer configured at start-up via `priv/signs.json`. See `Signs.Utilities.SourceConfig` for documentation of the format of the `signs.json` file.
+
+There are several "Engine" GenServers, responsible for maintaining various bits of data: alerts, predictions, signs-ui configuration, etc. Each sign queries the engines once per tick (roughly 1/sec) to see if its content needs to change.
+
+To prevent DOSing the ARINC head-end server, each sign does not send HTTP requests to it directly. Rather, there's a `MessageQueue` GenServer that the signs use to add their messages to. Separately, there are a number of `PaEss.HttpUpdater` GenServers that pull from the queue to send the messages. The number of updaters and the frequency with which they pull from the queue is fixed, in order to rate limit the requests to the head end server. The size of the MessageQueue is fixed to allow some buffering for hiccups in the network, while not being so large that very out of date messages end up getting sent to ARINC. The ARINC head-end server has a lot of cores, but is somewhat slow per-core, so we ended up with a largish number of updaters, each posting somewhat slowly.
```

Comment:
> each sign does not send HTTP requests to it directly

> that the signs use to add their messages to

Same comment as above, basically. It seems like we're saying "a sign" when we mean "a zone" throughout this doc, especially in the "utility modules" section. We could say that this is the convention we use, but in that case it'd help to call it out explicitly at the beginning.